Enhancements: Formula Card and Rule Action

lelemm opened the pull request:
Based on initial feedback, some fixes and enhancements to it:
- Added `balance` to rule action formula
- Added `INTEGER_TO_TRANSACTION` to rule action formula
- Fixed queries in Formula Card overflowing the container
- Clean up unnecessary code from `packages/loot-core/src/server/rules/action.ts`

lelemm commented:
@youngcw some fixes and enhancements
I wasn't able to reproduce the query range filter yet. I will try again later, but if you can review this one while that

youngcw commented:
> It worked for me @youngcw
> 
> <img alt="image" width="903" height="469" src="https://private-user-images.githubusercontent.com/15043428/504384868-52e84279-1afc-4537-bd28-4ddda89cabcd.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjExNjQ5NzgsIm5iZiI6MTc2MTE2NDY3OCwicGF0aCI6Ii8xNTA0MzQyOC81MDQzODQ4NjgtNTJlODQyNzktMWFmYy00NTM3LWJkMjgtNGRkZGE4OWNhYmNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDIyVDIwMjQzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0MmE3OTNjNzZkMjc2ODM3MWMzNmZiMDBiNDM2YWEzMDk5MjgzOWU3YTRkZmVlZmVlN2RlZjI4NGUwMTMyNDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vw1m1MX4V1Qaut7KyCUwlAiiWOD_sm37KeuApkeqjcA">
> Also, I didn't change how `balance` is generated/calculated. I just added the balance to the suggestions

did you try to use that rule in the account page?

Rule
<img width="887" height="270" alt="image" src="https://github.com/user-attachments/assets/8a13b9b1-a9c3-4234-9ca9-c2658958c269" />

After the rule runs
<img width="852" height="126" alt="image" src="https://github.com/user-attachments/assets/e832b7b6-f586-4089-9573-b9c5ff7eb410" />

lelemm commented:
> > It worked for me @youngcw
> > <img alt="image" width="903" height="469" src="https://private-user-images.githubusercontent.com/15043428/504384868-52e84279-1afc-4537-bd28-4ddda89cabcd.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjExNjQ5NzgsIm5iZiI6MTc2MTE2NDY3OCwicGF0aCI6Ii8xNTA0MzQyOC81MDQzODQ4NjgtNTJlODQyNzktMWFmYy00NTM3LWJkMjgtNGRkZGE4OWNhYmNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDIyVDIwMjQzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0MmE3OTNjNzZkMjc2ODM3MWMzNmZiMDBiNDM2YWEzMDk5MjgzOWU3YTRkZmVlZmVlN2RlZjI4NGUwMTMyNDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vw1m1MX4V1Qaut7KyCUwlAiiWOD_sm37KeuApkeqjcA">
> > Also, I didn't change how `balance` is generated/calculated. I just added the balance to the suggestions
> 
> did you try to use that rule in the account page?
> 
> Rule <img alt="image" width="887" height="270" src="https://private-user-images.githubusercontent.com/28542559/504394522-8a13b9b1-a9c3-4234-9ca9-c2658958c269.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjExNjU4NTEsIm5iZiI6MTc2MTE2NTU1MSwicGF0aCI6Ii8yODU0MjU1OS81MDQzOTQ1MjItOGExM2I5YjEtYTljMy00MjM0LTljYTktYzI2NTg5NThjMjY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDIyVDIwMzkxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlhODJhMzg3NTg0Y2NiM2JiZmE4OWVhOWIyNTQ3Y2E4MjE0NTdjYzg0ZDFmZjhiZGU0MjIxNjJjM2M5NDE3ODImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8bjFHvL08xQlLqddl-GCCu_v-tmmJ-xFkLmcbZ4S7LE">
> 
> After the rule runs <img alt="image" width="852" height="126" src="https://private-user-images.githubusercontent.com/28542559/504394342-e832b7b6-f586-4089-9573-b9c5ff7eb410.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjExNjU4NTEsIm5iZiI6MTc2MTE2NTU1MSwicGF0aCI6Ii8yODU0MjU1OS81MDQzOTQzNDItZTgzMmI3YjYtZjU4Ni00MDg5LTk1NzMtYjljNWZmN2ViNDEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDIyVDIwMzkxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUyYmMwOGJlYjJlZTljOWQ1ZjZjMGFlZTMwZTUxMzYzNmFlMGYzMWE5Zjc4ZGU3YjgyZDhjMWFkNTA2ODhmMWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Dq73xtKwN38MpH73aglviMQcupBXS9H9wZUoiWn0tp8">
> 
> Putting `{{ test }}` as a notes template doesn't work. Adding {{ mul 1 100 }} as an amount template does work. So some rule templates work but others don't.

what does `{{ test }}` do as template?

joel-jeremy commented:
Idea: what if we could also use these formulas to create custom data pills on the accounts page e.g. 

<img width="1614" height="294" alt="image" src="https://github.com/user-attachments/assets/a075e719-c768-4e4a-b285-58a101c96767" />

Maybe users can create a pill to easily see Food categories, sum of all transaction within specific date ranges (credit card statements), etc.

lelemm commented:
> Idea: what if we could also use these formulas to create custom data pills on the accounts page e.g.
> 
> <img alt="image" width="1614" height="294" src="https://private-user-images.githubusercontent.com/20313680/505012776-a075e719-c768-4e4a-b285-58a101c96767.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjEzMDk1NjAsIm5iZiI6MTc2MTMwOTI2MCwicGF0aCI6Ii8yMDMxMzY4MC81MDUwMTI3NzYtYTA3NWU3MTktYzc2OC00ZTRhLWIyODUtNThhMTAxYzk2NzY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDI0VDEyMzQyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdiODNmYTM0ODI4ZmY3Zjg4N2Q3MDY2ODMyZjRkMTk0YmViOTUxYWM2ZTFhZjNhY2RjMWZkMmY3ZTBjMWYwZDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4RYXfzXywrMNy2TdlQThzcVS49etTPQrE_MiafUaVrw">
> Maybe users can create a pill to easily see Food categories, sum of all transaction within specific date ranges (credit card statements), etc.

I like the idea.
Do you think pills per account? general pills for all accounts?
How do you see this being persisted? New table? new field on accounts table? synced_prefs?

lelemm commented:
@youngcw I was thinking here about your balance issue.
Are you comparing the balance with the account balance directly?

<img width="543" height="148" alt="image" src="https://github.com/user-attachments/assets/061f69e9-4656-46e3-9314-e2225fa0f508" />

https://actualbudget.org/docs/experimental/rule-templating/#variables

the balance from formula and template subtracts the current account value.

youngcw commented on packages/desktop-client/src/components/reports/reports/FormulaCard.tsx:
```diff
@@ -47,7 +47,7 @@ export function FormulaCard({
       ...Object.entries(themeColors).reduce(
         (acc, [key, value]) => {
-          acc[`theme.${key}`] = value;
+          acc[`theme_${key}`] = value;
```
It doesn't matter much, but is there a reason you are preferring "_" here instead of a "."?  I thought dot made more sense, but maybe thats just because Im used to programming syntax.

youngcw commented:
There is some visual strangeness now.  I don't get the tooltip of available functions anymore and the edit fields for the formula report and formulas in rules don't look to have the right colors.

lelemm commented:
> It would be really cool if there was also a `balanceOnDay()` function. That way you could make schedules that can predict a CC payment if you are paying the statement balance, not the full balance on the day of payment.

add these to the feedback issue please @youngcw . I will not be able to hit all the feedbacks in one PR.
https://github.com/actualbudget/actual/issues/3606

lelemm commented:
> Idea: what if we could also use these formulas to create custom data pills on the accounts page e.g.
> 
> <img alt="image" width="1614" height="294" src="https://private-user-images.githubusercontent.com/20313680/505012776-a075e719-c768-4e4a-b285-58a101c96767.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM4NzU5MDMsIm5iZiI6MTc2Mzg3NTYwMywicGF0aCI6Ii8yMDMxMzY4MC81MDUwMTI3NzYtYTA3NWU3MTktYzc2OC00ZTRhLWIyODUtNThhMTAxYzk2NzY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTIzVDA1MjY0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk3ZDVmNzgyY2UxN2VmYjUxMGQ5MGFmMGY4ZmMxMmM0ZThjMTliZTlhNDNkYmJmZWM0YWE1MzhhMTgxNzEzZTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.cE81McBRhmTPG2z_99Su9EY5Fpx99jHgrtU1AD0jFwY">
> Maybe users can create a pill to easily see Food categories, sum of all transaction within specific date ranges (credit card statements), etc.



> > > Idea: what if we could also use these formulas to create custom data pills on the accounts page e.g.
> > > <img alt="image" width="1614" height="294" src="https://private-user-images.githubusercontent.com/20313680/505012776-a075e719-c768-4e4a-b285-58a101c96767.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjEzMDk1NjAsIm5iZiI6MTc2MTMwOTI2MCwicGF0aCI6Ii8yMDMxMzY4MC81MDUwMTI3NzYtYTA3NWU3MTktYzc2OC00ZTRhLWIyODUtNThhMTAxYzk2NzY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDI0VDEyMzQyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdiODNmYTM0ODI4ZmY3Zjg4N2Q3MDY2ODMyZjRkMTk0YmViOTUxYWM2ZTFhZjNhY2RjMWZkMmY3ZTBjMWYwZDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4RYXfzXywrMNy2TdlQThzcVS49etTPQrE_MiafUaVrw">
> > > Maybe users can create a pill to easily see Food categories, sum of all transaction within specific date ranges (credit card statements), etc.
> > 
> > 
> > I like the idea. Do you think pills per account? general pills for all accounts? How do you see this being persisted? New table? new field on accounts table? synced_prefs?
> 
> Good questions. I think it can be done 2 ways:
> 
> 1. Reuse results of formula report cards
>    
>    * When creating a new pill, we just allow users to choose from existing formula cards and display the result of that report/card on the pill
> 2. New formulas table
>    
>    * Have a dedicated Formulas page where users can create and save a bunch of reusable formulas which can be then be referenced in formula report cards, account pills, rules, etc.
> 
> Parameterized formulas/queries would also be very handy e.g. `QUERY('Current month's Food spending', accountId)`

can you add these suggestions at https://github.com/actualbudget/actual/issues/3606 @joel-jeremy ?

Thanks

lelemm commented:
wrong link @youngcw and @joel-jeremy https://github.com/actualbudget/actual/issues/5949

lelemm commented:
I'm sticking this PR with only this feedbacks:

Added balance to rule action formula
Added INTEGER_TO_TRANSACTION to rule action formula
Fixed queries in Formula Card overflowing the container
Clean up unnecessary code from packages/loot-core/src/server/rules/action.ts

I will add create another PR later with more feedbacks
